piexif: 1.0.13 -> 1.1.2 (and bunch of other packages)

jluttine opened the pull request:
###### Motivation for this change

I just wanted to upgrade Piexif but Thumbor didn't have tests enabled so I didn't dare to upgrade without having tests passing successfully. Well, enabling the tests required quite a few new packages to be added...

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

jluttine committed:
pythonPackages.should-dsl: init at 2.1.2

jluttine committed:
pythonPackages.nose-pattern-exclude: init at 0.1.3

jluttine committed:
pythonPackages.nose-of-yeti: init at 1.8

jluttine committed:
pythonPackages.nose-focus: init at 0.1.3

jluttine committed:
pythonPackages.yanc: init at 0.3.3

jluttine committed:
pythonPackages.remotecv: init at 2.2.2

jluttine committed:
pythonPackages.pyres: init at 1.5

jluttine committed:
pythonPackages.thumbor: run tests

jluttine committed:
pythonPackages.pyssim: init at 0.4

jluttine committed:
pythonPackages.preggy: init at 1.4.2

jluttine committed:
pythonPackages.cairosvg1: init at 1.0.22

jluttine committed:
pythonPackages.piexif: 1.0.13 -> 1.1.2

veprbl reviewed (approved):
Looks great! This could use a rebase and reorder commits to enable thumbor tests after you add pyssim, preggy and cairosvg1.

veprbl commented:
cc @Ma27

worldofpeace commented on pkgs/development/python-modules/nose-focus/default.nix:
```diff
@@ -0,0 +1,35 @@
+, nose-pattern-exclude, which }:
+
+buildPythonPackage rec {
+  pname = "nose_focus";
```
```suggestion
  pname = "nose-focus";
```

jluttine replied:
Thanks, fixed now! I just used the "official" PyPI name here as I thought it should be used here, although I used the normalised name elsewhere.

worldofpeace commented on pkgs/development/python-modules/nose-of-yeti/default.nix:
```diff
@@ -0,0 +1,30 @@
+{ stdenv, buildPythonPackage, fetchFromGitHub, six, nose, fudge, should-dsl }:
+
+buildPythonPackage rec {
+  pname = "noseOfYeti";
```
```suggestion
  pname = "nose-of-yeti";
```

jluttine replied:
Thanks, fixed now! I just used the "official" PyPI name here as I thought it should be used here, although I used the normalised name elsewhere.

Ma27 commented:
First of all thank you @jluttine for the work!
As I was notified here (I guess because I maintain `thumbor`), I'd ask you to wait with merging until tomorrow, then I can have a look at this as well :)

worldofpeace commented on pkgs/development/python-modules/should-dsl/default.nix:
```diff
@@ -0,0 +1,21 @@
+{ stdenv, buildPythonPackage, fetchPypi }:
+
+buildPythonPackage rec {
+  pname = "should_dsl";
```
```suggestion
  pname = "should-dsl";
```

jluttine replied:
Thanks, fixed now! I just used the "official" PyPI name here as I thought it should be used here, although I used the normalised name elsewhere.

Ma27 commented on pkgs/development/python-modules/cairosvg/1_x.nix:
```diff
@@ -0,0 +1,34 @@
+  doCheck = false;
+
+  checkPhase = ''
+    FONTCONFIG_FILE=${fontconfig.out}/etc/fonts/fonts.conf nosetests .
```
in case the tests fail, we don't need a `checkPhase` here.

Ma27 replied:
that's even better! We don't want to pull check dependencies we don't need (yet) :+1:

jluttine replied:
I fixed this but now it's gone..

Ma27 commented on pkgs/development/python-modules/cairosvg/1_x.nix:
```diff
@@ -0,0 +1,34 @@
+  meta = with stdenv.lib; {
+    homepage = https://cairosvg.org;
+    license = licenses.lgpl3;
+    description = "SVG converter based on Cairo";
```
is it intentional that you didn't add yourself as maintainer here (IIRC we don't have a hard rule for that, but I prefer to merge packages to `master` that have a maintainer :))

jluttine replied:
I fixed this but now it's gone..

Ma27 commented on pkgs/development/python-modules/remotecv/install_requires.patch:
```diff
@@ -0,0 +1,16 @@
+-        "pyres>=1.5,<1.6",
+-        "Pillow>=4.3.0,<5.2.0",
++        "pyres>=1.5",
++        "Pillow>=4.3.0",
```
I'm not sure if we want version constraints without upper bounds here. When major bumps happen (and `remotecv` isn't compatible with them) we may end up having succeeding packages with broken build products.

I recommend something like `pyres>=1.5,<2.0`.

jluttine replied:
Thanks for the comment! I think I disagree though, if I'm understanding correctly. Please don't hesitate to disagree with me.

First, regarding remotecv itself: In my opinion using version upper bounds "just in case" is a bad practice. Remotecv needs to do a new release every time there's a new release of any of its dependencies it has upper bounded. If every package had that kind of policy, the whole Python ecosystem would be practically broken because of a need of recursive new releases. Of course, if the version is incompatible, then yes, needs an upper bound. Someone has written an issue about this: https://github.com/thumbor/remotecv/issues/15

Second, regarding our packaging: I'm not sure why we should replace the upper bounds written by Remotecv itself with something else instead of removing them. If the package creator has set 1.6 upper bound, why would we put 2.0 instead of just removing it? We have the check phase to make sure it works. If we set our own upper bound, we need to modify this patch whenever pyres or pillow is upgraded on NixOS. We aren't setting our own version upper bounds to any other packages just in case. That would cause recursive modification of package definitions whenever a dependency is upgraded. Or is there some particular reason that with remotecv and pyres/pillow this would be a good practice even if it wouldn't be good in general?

Ma27 replied:
> First, regarding remotecv itself: In my opinion using version upper bounds "just in case" is a bad practice. Remotecv needs to do a new release every time there's a new release of any of its dependencies it has upper bounded. If every package had that kind of policy, the whole Python ecosystem would be practically broken because of a need of recursive new releases. Of course, if the version is incompatible, then yes, needs an upper bound.

Your argument certainly applies for "minor" upper bounds (e.g. `>=1.2,<1.3`). But I'm talking about disallowing the next major version (e.g. `>=1.2,<2.0`). Major releases are expected to have API incompatibilities (as of SemVer), so I don't see why restricting them is an issue. I absolutely agree with you though that it's bad idea to have upper bounds as used in upstream `remotecv` though, in such a case you're absolutely right, one would have to update the package after each minor (non-breaking) release.

> Second, regarding our packaging: I'm not sure why we should replace the upper bounds written by Remotecv itself with something else instead of removing them. If the package creator has set 1.6 upper bound, why would we put 2.0 instead of just removing it? We have the check phase to make sure it works. If we set our own upper bound, we need to modify this patch whenever pyres or pillow is upgraded on NixOS. We aren't setting our own version upper bounds to any other packages just in case.

I think that restricting `2.0` is the most sensitive variant. Your patch allows e.g. `pyres` 3.0 ATM as well.

Let's say folks don't upgrade `pyres` in the next months and upstream releases 3.0. A bump happens in `nixpkgs` and `remotecv` still builds **although** it's incompatible with `pyres` 3.0 (e.g. since 3.0 support isn't released yet). Now we have a working build, but the result can be broken at runtime due to API changes.

I prefer to actually have breaking packages when doing a major bump (these can be covered easily with `nix-review`) rather than weird runtime failures. Also it's not clear if a simple update of the version constraint is sufficient, you may have to upgrade this package as well to get it working (or maintain two major versions of $package).

I hope that makes my concerns a bit clearer :)

jluttine replied:
I see your concern but I still slightly disagree: I think unit tests should check if it'll work runtime or not. Working build should thus mean working runtime compatibility with the dependencies. If unit tests don't capture runtime failures in case of dependency incompatibility, I think it's a problem in the unit tests. So, if `remotecv` builds and checks fine, then it should work runtime with the given dependencies. And this is all done automatically in nixpkgs, thus I don't see any reason to add this kind of manual restrictions unless there are some particular concerns on some package dependencies or the upstream package has explicitly made those upper bounds for some good reason.

`nox-review` will also build recursively and run check phases this so I don't see the problem with that either.. Is your concern that the unit test coverage on `remotecv` is bad and we should thus protect these particular dependency upgrades? For which other packages should we add this kind of upper bound restrictions too? But I can add that change if wanted, no problem, I just don't think it's good practice as far as I understand.

Ma27 replied:
Ah right that's clearly my mistake, I commented about some `checkPhase` things earlier and I had in mind that the patch was for this package.

That's clearly my mistake, sorry for the noise!

jluttine replied:
Ah, ok!

Ma27 commented on pkgs/development/python-modules/thumbor/default.nix:
```diff
@@ -27,18 +53,26 @@ buildPythonPackage rec {
+    pkgs.exiftool
+    pkgs.libjpeg
+    pkgs.ffmpeg
+    pkgs.gifsicle
```
AFAIK it's not needed to include `pkgs` into this expression, adding `exiftool` etc to the argument list should be sufficient

jluttine replied:
Ok! I thought it'd be more explicit which packages are Python packages and which are system packages. Because there are name collisions at least with `redis`. Or is there some other way to request for the system package `redis` instead of Python package `redis` as an argument? As I needed to use `pkgs` for that, I thought why not use the same for all system packages. But I can remove `pkgs.` from all other except `pkgs.redis` if that's better.

Ma27 replied:
I guess you're right. Some packages have Python bindings with the same name in the Python package set. Adding `pkgs` makes it clearer :+1:

Ma27 commented on pkgs/development/python-modules/yanc/default.nix:
```diff
@@ -0,0 +1,27 @@
+  version = "0.3.3";
+
+  # Tests fail on Python>=3.5. See: https://github.com/0compute/yanc/issues/10
+  disabled = !(pythonOlder "3.5");
```
does "tests fail" mean that the package is broken? Otherwise you could simply write `doCheck = pythonOlder "2.5"`.

jluttine replied:
I don't know. :slightly_smiling_face: I'm not familiar with the package but I would assume the package doesn't work as expected if the tests fail, thus I'd disable the package. I've made a bug report upstream (the link in the comment): https://github.com/0compute/yanc/issues/10

Ma27 reviewed (changes requested):
Thanks a lot for working hard on improving these python packages (and all the dependency issues in here). I added some minor comments (I'm also happy to discuss about).

I just tested `thumbor` locally and it seems to still work :)

jluttine commented:
@veprbl Thanks for the comments!

I tried rebasing on master but got some conflicts not related to this pull request so I didn't dare proceed solving those. This is now based on nixos-unstable.

GitHub shows the commits in weird order. Probably based on some timestamps. The commits are already in the order you are requesting.

Ma27 commented:
> I  tried rebasing on master but got some conflicts not related to this  pull request so I didn't dare proceed solving those. This is now based  on nixos-unstable. GitHub shows the commits in weird order. Probably based on some  timestamps. The commits are already in the order you are requesting.

Indeed, the GitHub timeline can be a bit confusing, especially when force-pushing onto your branch. The history of a branch is usually clearer when looking at the "Commits" tab.

I'll have a look at the conflict, but I won't push a resolution (yet).

Ma27 commented:
@jluttine it seems as you checked out your branch before the change(s) were merge that moved all remaining packages from `pkgs/top-level/python-packages.nix`into their own files.

From what I can see it's only one large conflict which I resolved locally. I'll test all the changes now and I'd ask for your permission to push this onto your branch (I can do this as NixOS maintainer, but in such a case I'd ask for permission first ^^)

jluttine commented:
@Ma27 Great, thanks! Feel free to push to this branch.

Ma27 commented:
btw I'm getting irreproducible build failures during the Redis tests, that are fixed when restarting the build.

Ma27 commented:
The problem causing the test failures on my checkout is the following commit: 094e64318801cb1ba86dfa244e7869bb4dc9bad9

I'll look into the upstream sources (reverting this locally fixed the `thumbor` build again. I'll force-push the fixed patches as soon as I've found a suitable solution, but please recheck if everything works for you as well ok @jluttine ?

Ma27 committed:
pythonPackages.pyres: ensure that redis 2.10.6 is used

Upstream's requirements.txt wrongly suggests that redis==3.x is support
(missing upper-bound constraint), but the tests break due to API
incompatibilities.

Ma27 commented:
@GrahamcOfBorg build pythonPackages.thumbor pythonPackages.piexif python3Packages.piexif

Ma27 commented:
Let's wait for ofBorg, after that the change should be fine :)

GrahamcOfBorg commented:
<!--REQUEST_ID=482549c6-5c47-4531-8cd2-8b3839c75cfb-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.49409&attempt_id=adbfdd84-9d4e-40b6-bcb8-504ee792e6a6)

Attempted: pythonPackages.thumbor, pythonPackages.piexif, python3Packages.piexif

<details><summary>Partial log (click to expand)</summary><p>

```
test_logger_should_not_be_null (tests.test_utils.UtilsTestCase) ... ok

----------------------------------------------------------------------
Ran 492 tests in 33.385s

OK (SKIP=1)
Joining threads....
/nix/store/yybv9iy8q83qjnddxmh5njlmvfyr986h-python2.7-thumbor-6.5.2
/nix/store/rldf72y0rrn5j51pldlmwn3liyv4rh67-python2.7-piexif-1.1.2
/nix/store/w6xxv2x02b529wghg61qzl7lmn6iqsrn-python3.7-piexif-1.1.2
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=482549c6-5c47-4531-8cd2-8b3839c75cfb-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.49409&attempt_id=22f724f0-ede0-472f-aec2-4a0aae4ff24b)

Attempted: pythonPackages.thumbor, pythonPackages.piexif, python3Packages.piexif

<details><summary>Partial log (click to expand)</summary><p>

```
test_logger_should_not_be_null (tests.test_utils.UtilsTestCase) ... ok

----------------------------------------------------------------------
Ran 492 tests in 48.339s

OK (SKIP=1)
Joining threads....
/nix/store/n7rn86k9f0gm21sbz1m4vphnvk645agn-python2.7-thumbor-6.5.2
/nix/store/37bb0ylbgzi3vv5kh0q996sbln9sxw6f-python2.7-piexif-1.1.2
/nix/store/ab2qml7ys48l29wbc4xjdgnk8pgllh01-python3.7-piexif-1.1.2
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=482549c6-5c47-4531-8cd2-8b3839c75cfb-->
Timed out, unknown build status on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.49409&attempt_id=8f7dc5c0-ed58-4db7-b74e-c5b9c54ae2a5)

Attempted: pythonPackages.thumbor, pythonPackages.piexif, python3Packages.piexif

<details><summary>Partial log (click to expand)</summary><p>

```
libtool: compile:  clang++ -DHAVE_CONFIG_H -I. -I.. -pthread -DHAVE_PTHREAD=1 -DHAVE_ZLIB=1 -Wall -Wno-sign-compare -O2 -g -std=c++11 -DNDEBUG -c google/protobuf/compiler/java/java_message_builder.cc -o google/protobuf/compiler/java/java_message_builder.o >/dev/null 2>&1
/nix/store/sv35yjk452mggg76bz94nppibj96a66h-bash-4.4-p23/bin/bash ../libtool  --tag=CXX   --mode=compile clang++ -DHAVE_CONFIG_H -I. -I..    -pthread -DHAVE_PTHREAD=1 -DHAVE_ZLIB=1 -Wall -Wno-sign-compare -O2 -g -std=c++11 -DNDEBUG -c -o google/protobuf/compiler/java/java_message_builder_lite.lo google/protobuf/compiler/java/java_message_builder_lite.cc
libtool: compile:  clang++ -DHAVE_CONFIG_H -I. -I.. -pthread -DHAVE_PTHREAD=1 -DHAVE_ZLIB=1 -Wall -Wno-sign-compare -O2 -g -std=c++11 -DNDEBUG -c google/protobuf/compiler/java/java_message_builder_lite.cc  -fno-common -DPIC -o google/protobuf/compiler/java/.libs/java_message_builder_lite.o
libtool: compile:  clang++ -DHAVE_CONFIG_H -I. -I.. -pthread -DHAVE_PTHREAD=1 -DHAVE_ZLIB=1 -Wall -Wno-sign-compare -O2 -g -std=c++11 -DNDEBUG -c google/protobuf/compiler/java/java_message_builder_lite.cc -o google/protobuf/compiler/java/java_message_builder_lite.o >/dev/null 2>&1
building of '/nix/store/w783xj47dpgvc1gaqj22izfh5cmka8gs-protobuf-3.6.1.drv' timed out after 1800 seconds
cannot build derivation '/nix/store/4jb0vlx0pkqzpbivc90avbssipprza3p-opencv-3.4.3.drv': 3 dependencies couldn't be built
building of '/nix/store/jl77c90fs3jraylwdcafn3g5703rlgvp-python2.7-pyres-1.5.drv' timed out after 1800 seconds
cannot build derivation '/nix/store/ilvjzrj8xswgkc4ipr5zqam32xhs904k-python2.7-remotecv-2.2.2.drv': 3 dependencies couldn't be built
cannot build derivation '/nix/store/cndjzfd5rrrp3z0nnrhbzl3ib8hsfx5k-python2.7-thumbor-6.5.2.drv': 4 dependencies couldn't be built
error: build of '/nix/store/cndjzfd5rrrp3z0nnrhbzl3ib8hsfx5k-python2.7-thumbor-6.5.2.drv' failed
```
</p></details>

Ma27 reviewed (approved):
The change itself seems fine now, thanks a lot! :+1: 

I've observed several Darwin issues in the last days on Borg, it's possible that the Darwin package set isn't completely built, thus the timeout.

I'll wait with merging until @jluttine had a look at the new diff (after rebasing onto master and resolving the fairly big conflict in `pkgs/top-level/python-packages.nix` and fixing `pyres` to build on current master)

jluttine commented:
@Ma27 Looks like all my review fixes are gone.

Ma27 commented:
reapplied review comments and rebuilt packages, thanks! :)

jluttine commented:
Looks good. I ran `nix-shell -p pythonPackages.thumbor` and it successfully opened the shell. So I guess it's ok then. :slightly_smiling_face: I didn't run nox-review now.

Ma27 commented:
The only failing package when running `nix-review pr 49409` is `python37Packages.Nikola`, but this is failing on master as well: https://hydra.nixos.org/job/nixos/trunk-combined/nixpkgs.python37Packages.Nikola.x86_64-linux/all

I checked it locally and it doesn't seem to be related to your change (it wants `piexif>=1.0.3`, so your `piexif` bump won't break the installation)

Ma27 commented:
@jluttine thanks!
